feat: add themes supports

SupergroverNN opened the pull request:
Добавлен функционал изменения темы + сохранение настроек вибратора и звука (не вызывается сохранение, если звук отключается из-за неактивности окна)

SupergroverNN committed:
feat: add themes supports

ValeriaKoliagina commented on src/components/menu/MenuItem.tsx:
```diff
@@ -24,6 +23,17 @@ const MenuItem: FC<IProps> = ({ item, modifier, isMobile, handleAction }): React
     }
   };
 
+  if (title === 'theme' && typeof modifier[title] === 'string') {
+    const mode = modifier[title].toString();
```
а зачем нам приводить к стринге, если выше есть проверка, что это должно быть стринга?

SupergroverNN replied:
согласен, выглядит странно, но тс ругался почему то на то что если он будет булеан то не сработает .toUpperCase()

kotosha-real replied:
Лучше убрать тему из модификаторов и передавать отдельно. Либо точнее затипизировать `modifier`, то есть указать, что там не просто ключи — строки, а значения — строки или булеаны, а перечислить конкретные свойства. Правда, после того придётся делать проверки на `title` рядом с проверкой на `action`, иначе будет примерно такая ошибка:
```
Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ sound: boolean; vibration: boolean; theme: string; }'.
```

ValeriaKoliagina commented:
а почему у нас теперь ошибок не видно?
![image](https://user-images.githubusercontent.com/32342424/129260908-8f780793-114e-4e54-9cc9-2ccefd8e811a.png)

ValeriaKoliagina commented:
На ховер слеш остался фиолетовым)))
![image](https://user-images.githubusercontent.com/32342424/129261106-1c30b2e3-c88e-4b96-8649-ed0a817624ab.png)
и на логине на основной теме тоже теперь красный фон, его раньше не было, и тоже ошибок не видно
![image](https://user-images.githubusercontent.com/32342424/129261658-45515c75-f1b3-49c8-bd0b-6810aa30d74a.png)

ValeriaKoliagina commented:
В остальном все круто! Здорово!)

SupergroverNN commented:
> а почему у нас теперь ошибок не видно?
> ![image](https://user-images.githubusercontent.com/32342424/129260908-8f780793-114e-4e54-9cc9-2ccefd8e811a.png)

fixed

SupergroverNN commented:
> На ховер слеш остался фиолетовым)))
> ![image](https://user-images.githubusercontent.com/32342424/129261106-1c30b2e3-c88e-4b96-8649-ed0a817624ab.png)
fixed)

> и на логине на основной теме тоже теперь красный фон, его раньше не было, и тоже ошибок не видно
> ![image](https://user-images.githubusercontent.com/32342424/129261658-45515c75-f1b3-49c8-bd0b-6810aa30d74a.png)
а это, как выяснилось, нотификация запорола своим отдельным стилем .error фон для инпута с ошибкой =( 
fixed

SupergroverNN commented:
А еще такую штуку проверьте плизз, похоже починилась на этот раз.

Было так (у меня по крайней мере)
![before](https://user-images.githubusercontent.com/40541530/129297131-195afe64-44e4-4371-900b-0aa92f366bb8.jpg)

А теперь вот так (опять же у меня)
![after](https://user-images.githubusercontent.com/40541530/129297215-8a5794e7-8452-4251-b734-ddc1654df214.jpg)

SupergroverNN committed:
fix: PR issues

SupergroverNN committed:
Merge branch 'main' into feat-add-themes-support

StopNGo commented on src/style/common.scss:
```diff
@@ -2,7 +2,7 @@ html,
 #root {
   width: 100%;
-  height: 100%;
+  min-height: 100%;
```
Привет!
Теперь из-за этого у нас липнут окошки к верху. См. тут и ниже, как пока лучше временно поменять.
```suggestion
  height: 100%;
```

```
body {
  background: url('images/main_bg.jpg') no-repeat;
  background-attachment: fixed;
  background-size: cover;
}
```

```
.main {
  ...
  height: 100%;
  ...
}
```

SupergroverNN committed:
fix: styles

kotosha-real commented on src/components/gameCanvas/GameCanvas.tsx:
```diff
@@ -107,10 +110,19 @@ const GameCanvas: FC<IGameCanvas> = (
+      case MENU_ACTIONS.GAME_THEME_SWITCH:
+      {
+        const { theme } = settings;
+        const nextTheme = theme === 'primary' ? 'secondary' : 'primary';
```
❗️
Давай вынесем названия тем в константы

SupergroverNN replied:
fixed

kotosha-real commented on src/components/gameCanvas/GameCanvas.tsx:
```diff
@@ -27,12 +29,13 @@ interface IGameCanvas {
 }
 
 const GameCanvas: FC<IGameCanvas> = (
-  { className, isSoundOn, isVibrationOn, menuAction, resetMenuAction }
+  { className, isSoundOn, isVibrationOn, settings, menuAction, resetMenuAction }
 ): ReactElement => {
```
🧐
Я уже писал в другом пре, кажется, у Макса. Если деструктурируемых пропсов много, то лучше делать это внутри тела функции. Так объявление функции будет занимать одну строчку вместо нескольких, будет читабельнее и аккуратнее

SupergroverNN replied:
fixed

kotosha-real commented on src/components/menu/Menu.tsx:
```diff
@@ -20,7 +20,7 @@ interface IMenu {
   isWithTitle?: boolean;
   className?: string;
-  modifier: { [key: string]: boolean }
+  modifier: { [key: string]: boolean | string }
```
❗️
Подобный подход будет мешать тебе типизировать значения в этом объекте, всегда придётся прибегать к кастингу типов. У нас есть вполне понятные значения, которые всегда лежат в `modifier`, давай перечислим их вместо абстрактных строкоключей. Обрати внимание, что некоторые проверки могут из-за этого побиться, потому что есть проверки, которые подразумевают, что `title` (свойство, по котором ты достаёшь значения из `modifier`) — это просто строка, а не конкретные значения

kotosha-real replied:
❗️
Второй коммент туда же — такой же `modifier` есть и в `MenuItem`, давай определим его в одном месте, а в интерфейсах компонентов затипизируем вот так:
```suggestion
  modifier: TModifier
```

kotosha-real commented on src/components/menu/Menu.tsx:
```diff
@@ -35,13 +35,14 @@ const Menu: FC<IMenu> = ({
   handleAction
 }): ReactElement => {
   let isMobile = false;
+  const theme = modifier['theme'];
```
🧐
Не уверен, что `modifier` — лучшее место для хранения темы, кажется, что лучше её передавать отдельно. Это не призыв к действию, если есть резон так сделать, то можно оставить)

SupergroverNN replied:
Тут логически, получается что это модификатор, так как это пункт меню с модификатором и логика соответствующая )
Так что наверное имеет смысл оставить текущий подход )

kotosha-real commented on src/components/menu/MenuItem.tsx:
```diff
@@ -2,17 +2,16 @@ import React, { FC, ReactElement } from 'react';
 interface IProps {
   item: TMenuItem;
-  modifier:  { [key: string]: boolean };
+  modifier: { [key: string]: boolean | string };
```
❗️
Писал о типизации и DRY выше

SupergroverNN replied:
завел интерфейс для modifier

kotosha-real commented on src/pages/leaderboard/Leaderboard.tsx:
```diff
@@ -29,6 +29,7 @@ type LeaderState = ILeaders[] | []
 const Leaderboard: FC = (): ReactElement => {
   const dispatch = useDispatch();
   const { id: userId } = useSelector(userUserDataSelector);
+  const theme = useSelector(userSettingSelector).theme;
```
🧐
```suggestion
  const { theme } = useSelector(userSettingSelector);
```

SupergroverNN replied:
ой да)

SupergroverNN replied:
fixed

kotosha-real commented on src/pages/forum/Forum.tsx:
```diff
@@ -29,6 +30,7 @@ type TopicState = ITopic[] | []
 const Forum: FC = (): ReactElement => {
   const dispatch = useDispatch();
   const { id } = useSelector(userUserDataSelector);
+  const theme = useSelector(userSettingSelector).theme;
```
🧐
```suggestion
  const { theme } = useSelector(userSettingSelector);
```

SupergroverNN replied:
fixed

kotosha-real commented on src/pages/game/Game.tsx:
```diff
@@ -29,6 +30,8 @@ const Game: FC = (): ReactElement => {
   const isVibrationOn = useSelector(gameIsVibrationOnSelector);
```
❓
Есть смысл хранить это в `game`-редюсере? Теперь эти данные хранятся в базе, почему бы не юзать их целиком из `userSetting`?

SupergroverNN replied:
тут, соглашусь, вибрацию можно хранить в сеттингах, но звук придется хранить и там и там, так как при неактивности окна должно меняться локальное состояние звука, а не глобальное.

SupergroverNN replied:
fixed

kotosha-real reviewed (approved):
🕵🏻‍♂️
Отличная работа по довольно сложной фиче!
Есть комменты и вопросы, но ввиду отсутствия времени я не стал помечать их как обязательные к исправлению, но было бы круто по ним подумать/пообщаться. Особенно про необходимость использования состояний вибрации/звука из `game`-редюсера

SupergroverNN committed:
fix: PR issues 3

SupergroverNN committed:
Merge branch 'main' into feat-add-themes-support
